### PR TITLE
Run RESTCONF/NETCONF randomly over a testsuite

### PR DESCRIPTION
**Pull request by mattiaswal**
## Description
* Implement full netconf  compatibility when using restconf
* A lot of refactor in netconf to be able to include restconf in tests
* testing.md updated
* Add upstream bugfix for finit, rousette and sysrepo-cpp
## Checklist

Tick *relevant* boxes, this PR is-a or has-a:

- [ ] Bugfix
  - [ ] Regression tests
  - [ ] ChangeLog updates (for next release)
- [x] Feature
  - [ ] YANG model change => revision updated?
  - [x] Regression tests added?
  - [ ] ChangeLog updates (for next release)
  - [ ] Documentation added?
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (please detail in commit messages)
- [ ] Build related changes
- [x] Documentation content changes
- [ ] Other (please describe):

### CODE REVIEW

**Comment by mattiaswal** [comment]
RESTCONF run on hardware:
```
s Execution
`-- s 0001-all.yaml
    |-- o 0002-reproducible.py
    |-- o 0003-wait.py
    |-- o 0004-Misc tests
    |   `-- o 0005-operational_all.py
    |-- o 0006-ietf-system
    |   |-- o 0007-hostname.py
    |   |-- o 0008-add_delete_user.py
    |   |-- o 0009-user_admin.py
    |   |-- o 0010-timezone.py
    |   `-- o 0011-timezone_utc_offset.py
    |-- o 0012-ietf-interfaces
    |   |-- o 0013-vlan_ping.py
    |   |-- o 0014-ipv4_address.py
    |   |-- o 0015-iface_phys_address.py
    |   |-- o 0016-iface_status.py
    |   `-- o 0017-routing_basic.py
    |-- o 0018-infix-interfaces
    |   |-- o 0019-bridge_basic.py
    |   |-- o 0020-bridge_veth.py
    |   |-- o 0021-dual_bridge.py
    |   |-- o 0022-bridge_vlan.py
    |   |-- o 0023-ipv4_autoconf.py
    |   |-- o 0024-bridge_fwd_sgl_dut.py
    |   |-- o 0025-bridge_fwd_dual_dut.py
    |   |-- o 0026-bridge_vlan_separation.py
    |   |-- o 0027-igmp_basic.py
    |   |-- o 0028-igmp_vlan.py
    |   `-- o 0029-static_multicast_filters.py
    |-- s 0030-ietf-routing
    |   |-- o 0031-static_routing.py
    |   |-- o 0032-ospf_basic.py
    |   |-- o 0033-ospf_unnumbered_interface.py
    |   `-- s 0034-ospf_multiarea.py
    |-- o 0035-infix-containers
    |   |-- o 0036-container_basic.py
    |   |-- o 0037-container_bridge.py
    |   |-- o 0038-container_phys.py
    |   `-- o 0039-container_veth.py
    |-- o 0040-infix-dhcp
    |   |-- o 0041-dhcp_basic.py
    |   |-- o 0042-dhcp_router.py
    |   `-- o 0043-dhcp_routes.py
    `-- o 0044-ietf-hardware
        `-- o 0045-usb.py
```

**Comment by mattiaswal** [comment]
NETCONF run on hardware:
```
s Execution
`-- s 0001-all.yaml
    |-- o 0002-reproducible.py
    |-- o 0003-wait.py
    |-- o 0004-Misc tests
    |   `-- o 0005-operational_all.py
    |-- o 0006-ietf-system
    |   |-- o 0007-hostname.py
    |   |-- o 0008-add_delete_user.py
    |   |-- o 0009-user_admin.py
    |   |-- o 0010-timezone.py
    |   `-- o 0011-timezone_utc_offset.py
    |-- o 0012-ietf-interfaces
    |   |-- o 0013-vlan_ping.py
    |   |-- o 0014-ipv4_address.py
    |   |-- o 0015-iface_phys_address.py
    |   |-- o 0016-iface_status.py
    |   `-- o 0017-routing_basic.py
    |-- o 0018-infix-interfaces
    |   |-- o 0019-bridge_basic.py
    |   |-- o 0020-bridge_veth.py
    |   |-- o 0021-dual_bridge.py
    |   |-- o 0022-bridge_vlan.py
    |   |-- o 0023-ipv4_autoconf.py
    |   |-- o 0024-bridge_fwd_sgl_dut.py
    |   |-- o 0025-bridge_fwd_dual_dut.py
    |   |-- o 0026-bridge_vlan_separation.py
    |   |-- o 0027-igmp_basic.py
    |   |-- o 0028-igmp_vlan.py
    |   `-- o 0029-static_multicast_filters.py
    |-- s 0030-ietf-routing
    |   |-- o 0031-static_routing.py
    |   |-- o 0032-ospf_basic.py
    |   |-- o 0033-ospf_unnumbered_interface.py
    |   `-- s 0034-ospf_multiarea.py
    |-- o 0035-infix-containers
    |   |-- o 0036-container_basic.py
    |   |-- o 0037-container_bridge.py
    |   |-- o 0038-container_phys.py
    |   `-- o 0039-container_veth.py
    |-- o 0040-infix-dhcp
    |   |-- o 0041-dhcp_basic.py
    |   |-- o 0042-dhcp_router.py
    |   `-- o 0043-dhcp_routes.py
    `-- o 0044-ietf-hardware
        `-- o 0045-usb.py
```

**Comment by troglobit** [comment]
Looks good after that last couple of commits! :+1:

**Comment by troglobit** [comment]
My PRs have now been merged to main, so you can start your rebase work.  Sorry again for all the conflicts!

**Review by troglobit — APPROVED** [review]
Nice work, took a while to get there, but the end result is really great!

**Review by wkz — APPROVED** [review]
I'm approving this, because I'm not sure now much of my feedback is reasonable to address now, but please have a think about it.

Great to finally get to protocol parity on the test side of things! :clap: :clap:

File: test/infamy/netconf.py [hunk 1]
```diff
@@ -410,6 +392,17 @@ def get_iface_xpath(self, iface, path=None):
...
+    def get_iface(self, iface):
+        """Fetch target dict for iface and extract param from JSON"""
+        content = self.get_data(self.get_iface_xpath(iface))
+        interface=content.get("interfaces", {}).get("interface", None)
+
+        if interface is None:
+            return None
+
+        # This really not required, it is due a bug in rousette.
+        return interface[iface]
```
[the remark this answers is not in the document]
Comment (mattiaswal) [reply test/infamy/netconf.py:405]: This *should* really not be required :)

**Comment by mattiaswal** [comment]
Still passing on hardware:
```
s Execution
`-- s 0001-all.yaml
    |-- o 0002-reproducible.py
    |-- o 0003-wait.py
    |-- o 0004-Misc tests
    |   `-- o 0005-operational_all.py
    |-- o 0006-ietf-system
    |   |-- o 0007-hostname.py
    |   |-- o 0008-add_delete_user.py
    |   |-- o 0009-user_admin.py
    |   |-- o 0010-timezone.py
    |   `-- o 0011-timezone_utc_offset.py
    |-- o 0012-ietf-interfaces
    |   |-- o 0013-vlan_ping.py
    |   |-- o 0014-ipv4_address.py
    |   |-- o 0015-ipv6_address.py
    |   |-- o 0016-iface_phys_address.py
    |   |-- o 0017-iface_status.py
    |   `-- o 0018-routing_basic.py
    |-- o 0019-infix-interfaces
    |   |-- o 0020-bridge_basic.py
    |   |-- o 0021-bridge_veth.py
    |   |-- o 0022-dual_bridge.py
    |   |-- o 0023-bridge_vlan.py
    |   |-- o 0024-ipv4_autoconf.py
    |   |-- o 0025-bridge_fwd_sgl_dut.py
    |   |-- o 0026-bridge_fwd_dual_dut.py
    |   |-- o 0027-bridge_vlan_separation.py
    |   |-- o 0028-igmp_basic.py
    |   |-- o 0029-igmp_vlan.py
    |   |-- o 0030-static_multicast_filters.py
    |   `-- o 0031-vlan_qos.py
    |-- s 0032-ietf-routing
    |   |-- o 0033-static_routing.py
    |   |-- o 0034-ospf_basic.py
    |   |-- o 0035-ospf_unnumbered_interface.py
    |   `-- s 0036-ospf_multiarea.py
    |-- o 0037-infix-containers
    |   |-- o 0038-container_basic.py
    |   |-- o 0039-container_bridge.py
    |   |-- o 0040-container_phys.py
    |   `-- o 0041-container_veth.py
    |-- o 0042-infix-dhcp
    |   |-- o 0043-dhcp_basic.py
    |   |-- o 0044-dhcp_router.py
    |   `-- o 0045-dhcp_routes.py
    `-- o 0046-ietf-hardware
        `-- o 0047-usb.py
13:01:38 infamy0:test #  
```

**Comment by troglobit** [comment]
Amazing work on this, thank you for sticking with it!! :heart:

**Pull request merged**